Voja

neworderofjamie opened the pull request.

Commit:
Groundwork for PES support
* Routing of modulatory connections and learnt connections
* Re-combining of both in LIF

Commit:
Attempt at building a PES region - unexpected core failure

Commit:
Merge remote-tracking branch 'origin/processes' into pes

Commit:
C changes
* Removed unused struct
* Flipped sign of error signal

Commit:
Restored multiplcation of learning rate by dt - That sure as hell doesn't help the situation

Commit:
Somewhat hacky attempt at long fract learning rate

neworderofjamie committed:
Somewhat hacky attempt at long fract learning rate (reverted from commit 76f94efd2bbb144542930b152edb0b8706b1959a)

neworderofjamie committed:
Merge remote-tracking branch 'origin/master' into pes

Conflicts:
	nengo_spinnaker/operators/lif.py

neworderofjamie committed:
We really don't want to be building thumb (the combination of settings also caused internal compiler error for me)

neworderofjamie committed:
Removed debug printing, fixed scale

neworderofjamie committed:
Removed more prints, added some comments

neworderofjamie committed:
Make sure learnt decoders aren't getting minimised and W.I.P. regression test

neworderofjamie committed:
pep8

neworderofjamie committed:
more pep8

neworderofjamie committed:
a test for PESRegion

neworderofjamie committed:
test for connections that sink in learning rules

Commit:
magic

Commit:
less magic, more spelling

Commit:
duh

Commit:
Formalised regression test and added new test for ensemble sources

Commit:
Filtered activity for learning
* Learning rules which require filtered activity with same time constant are gathered
* Filter is updated when neurons spike
* PES is evaluated for all neurons not just those that spike using filtered activity

Commit:
Voja connection routing
* Using error_type looks incorrect - switched to examining of pre_modifies and post_modifies
* Reroutes voja 'error' connections to post-synaptic ensemble
* Sinks voja learnt connections to unique ports identified by learning rule

Commit:
Start work on C-side of Voja

Commit:
Wooooo - finally found and fixed bug

Commit:
pep8

Commit:
Merge remote-tracking branch 'origin/pes' into voja

Conflicts:
	nengo_spinnaker/operators/lif.py

Commit:
Rethought port usage - Now only modulatory connections use learning rules as ports
* Connections with decoder learning rules are sourced from learnt port
* Connections with encoder learning rules are sunk into learnt port
* PES re-implemented in this framework
* Voja seems more achievable in this framework!

Commit:
More Voja
* Region reshuffle to make space somewhere for Voja
* C code building
* More Python work

Commit:
Merge remote-tracking branch 'origin/master' into voja

Commit:
* Commented out non-implemented bits of Voja
* New executables

Commit:
Both Voja and PES can operate on both filtered and non-filtered neural activity

Commit:
Encoder duplication groundwork
* More region reshuffling
* Encoders are duplicated for each Voja rule
* Encoder region is prepended with full width (only required for copying)
* Yet Another Filter (without accumulator) for inputs with learnt encoders - inputs here are assumed to have one-to-one mapping with duplicated encoder slices and these are combined together in ensemble_update
* Also did some c99ification of ensemble-update (making it clear scope of variables is not global can only help the compiler make stack vs register decisions) and a bit of general tidying

Commit:
Voja-on!
* Fixed warnings
* Enabled Voja

Commit:
Non-filtered voja implementation

Commit:
Renamed recording to make it spike-specific prior to implementation of recording_encoder

Commit:
Such a name pedant, record_spikes is maybe better

Commit:
Framework in place for encoder probing - Some bugs remain..

Commit:
Voja works!
* A little bit rough around the edges, but functioning
* Additional scale parameter added to end of Voja region and included in update equation
* Fixed sign of error signal
* Fixed lots of scary bugs I caused

Commit:
Final Voja tweaks
* Removed gain/radius array from voja and replacing with a 1/radius scalar which gets multiplied by existing gain
* Renamed inhib_gain to gain to make more generic

Commit:
voja branch requirements

Commit:
* Fixed broken test
* Added tests for new encoder-learning builder logic

Commit:
Fixed LIF tests

Commit:
pep8

mundya committed:
Remove unused filter code

mundya committed:
Rewrite of filter functionality

This commit makes some major changes to the ways filters are handled in
C.  Please refer to the code for detailed comments.  In brief:

  - A filter is a function which accepts a dimensionality, an input
    vector, an output vector and some filter-specific state.
  - Instances of filters are instantiated with function pointers which
    point at the function which should be used to evaluate them

The way the data is packed to be transmitted to the C has changed to
represent these changes. Each filter entry (as written out by the
Python) contains at least four words. The first word indicates the
number of extra words that are the parameters for the filter; the second
word indexes an array of function pointers to find the initialisation
method for the given filter; the third word indicates the "width" of the
filter (the number of dimensions in the input and output vector) and the
fourth word contains flags. Additional words contain filter-specific
parameters (e.g., coefficients, indication of order, etc.).

An additional test has been included to check that the filters work
correctly in C. **No new filter types have been implemented in this
commit.**

mundya committed:
Fix malloc and mask initialisation

 - Latching masks should be `0x00000000` not `0xffffffff`
 - Wrong size struct was malloced

mundya committed:
Add debug messages

Commit:
Merge remote-tracking branch 'origin/master' into voja

Conflicts:
	nengo_spinnaker/binaries/nengo_ensemble.aplx
	nengo_spinnaker/operators/lif.py
	spinnaker_components/ensemble/ensemble.h
	spinnaker_components/ensemble/ensemble_harness.c
	spinnaker_components/ensemble/ensemble_update.c

Commit:
Profiler - C-side
* Re-integrated profiler (newer version with support for disabling FIQ and IRQ if required)
* Updated profiler tags to cover PES, Voja and filtered output
* Updated Makefiles to (optionally) build targets for each executable with profiling enabled
* Initialised num_profiler_samples to 0 from python
* Hopefully fixed SystemRegion test

mundya committed:
Add LinearFilter support

 - Fix point rounding has a horrific effect on performance, this should
   be addressed in the future.

mundya committed:
Slight test fix

Needs better tolerances for the 2nd order system. Not sure quite what to
do with that.

Commit:
Approximately working profiler
* New config options for ensemble to enable profiling and manually set number of entries to record (something seems a little wrong here)
* Profiling region with helper method to read back profiling data (allows it to be used in any operator that supports profiling)
* Helper function to display summary of profile
* Nengo SpiNNaker simulator has profiler_data to hold data returned by any profiling activity

Commit:
Small fixes
* Profiler writes count in words not samples
* Timer tag should be opened after exit logic so a tag isn't leaked

Commit:
PEP8

Commit:
Merge remote-tracking branch 'origin/master' into profiling

Commit:
More PEP8

Commit:
Use warnings for warnings, don't upset the new-fangled world of Python 3

Commit:
weeping

Commit:
Support for basic CSV output from profiler

Commit:
pep8 and I are not having a good day

mundya commented:
I look forward to extracting profiling from this :/

mundya commented on nengo_spinnaker/builder/ensemble.py:
```diff
@@ -226,3 +301,29 @@ def build_neurons_probe(model, probe):
+
+
+@Model.probe_builders.register(nengo.connection.LearningRule)
+def build_voja_probe(model, probe):
```
Could this be `build_learnt_connection_probe` or similar?

mundya commented on spinnaker_components/Makefile.common:
```diff
@@ -163,7 +172,7 @@ endif
 
 
-# Build the ELF file
+# Build the ELF filBUILDDIRe
```
Huh?

neworderofjamie replied:
I think that's probably a good old fashioned copy-paste flail :)

mundya commented on spinnaker_components/ensemble/ensemble.h:
```diff
@@ -34,57 +34,85 @@
+extern input_filter_t g_input_inhibitory;
+
+// Input filters and buffers for modulatory signals. Their
+// outputs are left seperate for use by learning rules
```
sp. separate

mundya commented on spinnaker_components/ensemble/ensemble.h:
```diff
@@ -34,57 +34,85 @@
+extern input_filter_t g_input_modulatory;
+
+// Input filters and buffers for signals to be encoded by learnt encoders.
+// Each output is encoded by a seperate encoder so these are also left seperate
```
sp. separate

mundya commented on spinnaker_components/ensemble/ensemble_filtered_activity.c:
```diff
@@ -0,0 +1,83 @@
+    MALLOC_FAIL_FALSE(g_filtered_activities,
+                      g_num_activity_filters * sizeof(value_t*));
+
+    // Copy propogators from region into new array
```
sp.

mundya commented on spinnaker_components/ensemble/ensemble_filtered_activity.c:
```diff
@@ -0,0 +1,83 @@
+  // Loop through filters
+  for(uint f = 0; f < g_num_activity_filters; f++)
+  {
+    // Loop through neurons and apply propogators
```
sp.

mundya commented on spinnaker_components/ensemble/ensemble_pes.c:
```diff
@@ -60,9 +47,47 @@ bool get_pes(address_t address)
+    // Extract filtered error signal vector indexed by learning rule
+    const pes_parameters_t *parameters = &g_pes_learning_rules[l];
+
+    // If this learning rule operates on filtered activity and should, therefore be updated here
```
Say what now?

mundya commented on spinnaker_components/ensemble/ensemble_pes.h:
```diff
@@ -53,20 +57,21 @@ static inline void pes_neuron_spiked(uint n)
   // Loop through all the learning rules
   for(uint32_t l = 0; l < g_num_pes_learning_rules; l++)
   {
+    // If this learning rule operates on un-filtered activity and should, therefore be updated here
```
Huh?

neworderofjamie replied:
Sure, am in all day, probably staring at progress bars a lot!

mundya commented:
This might be going through in person if you've time? Sorry

mundya committed:
FIXUP Suggestions by @neworderofjamie

mundya committed:
FIXUP Ravel `a` and `b`

If SpiNNaker had cache this would be good... Another @neworderofjamie
suggestion.

mundya committed:
FIXUP Final neatening

Further suggestions by @neworderofjamie.

mundya committed:
FIXUP Redundant struct removed

Commit:
Spelling

Commit:
Merge remote-tracking branch 'origin/new-filters' into profiling

Conflicts:
	nengo_spinnaker/binaries/nengo_ensemble.aplx
	nengo_spinnaker/binaries/nengo_filter.aplx
	nengo_spinnaker/binaries/nengo_mc_player.aplx
	nengo_spinnaker/binaries/nengo_rx.aplx
	nengo_spinnaker/binaries/nengo_tx.aplx
	nengo_spinnaker/binaries/nengo_value_sink.aplx
	nengo_spinnaker/binaries/nengo_value_source.aplx
	spinnaker_components/common/input_filter.c
	spinnaker_components/common/input_filter.h
	spinnaker_components/ensemble/Makefile
	spinnaker_components/ensemble/ensemble.h
	spinnaker_components/ensemble/ensemble_harness.c
	spinnaker_components/ensemble/ensemble_main.c
	spinnaker_components/ensemble/ensemble_pes.h
	spinnaker_components/ensemble/ensemble_update.c
	spinnaker_components/filter/Makefile
	spinnaker_components/sdp_tx/Makefile
	spinnaker_components/value_sink/Makefile

Commit:
Out of boolean algebra fail

mundya commented:
This now includes #43 and #37

mundya commented:
So this is waiting on ensuring that everything still runs in PES and Voja are not present/the expected version?

mundya commented:
Closing this PR because the long suffering @neworderofjamie is working on another implementation which builds on #87.
